### ISSUE TITLE
AsyncApi -> Boilerplate generation of message and handlers

### ISSUE THREAD

**Issue by SbiCA** [rebus-org/Rebus#1017]
Hi @mookid8000 ,

I recently stumbled across https://www.asyncapi.com/ not sure if rebus is the right place, but to me it seems to a nice way of describing overall interaction and maybe there would be the potential to either generate boilerplate code or generate the docs similar to what swashbuckle,nswag allow you to do with OpenAPI

Interested to hear you're thoughts 😃 

Cheers

**Comment by SbiCA** [comment]
Fyi I saw there is already something in that regard https://github.com/neuroglia-io/AsyncApi but of course it's not rebus ;)

**Comment by mookid8000** [comment]
Hi @SbiCA , AsyncAPI definitely looks interesting 🙂 

Did you have any specific usage scenarios in mind?

One possible usage scenario I could imagine, was to help with enforcing message correspondence protocols by enabling that a Rebus instance could be configured with its intended API in the configuration, and then it would make the AsyncAPI document available somehow (e.g. it could be posted to [Fleet Manager](https://pro.rebus.fm/#fleet-manager) if one were using that).

I imagine pulling in a NuGet package called Rebus.AsyncApi, which would give me an `AsyncApi` extension method under `Options`, which would enable describing the service metadata and the intended contract, e.g. like this:

```csharp
services.AddRebus(
    configure => configure
        .Transport(t => t.UseAzureServiceBus(conn, "order_processor"))
        .Options(o => o.AsyncApi(api => {
            api.Title = "Order Processor";
            api.Version = new Version(2, 1, 0);
            api.Description = "Processes orders as they come in";

            // we handle commands of this type
            api.Handles<ProcessOrder>();

            // we publish events of this type
            api.Publishes<OrderProcessed>();

            // we subscribe to these events
            api.SubscribesTo<ShippingInfoAddedToOrder>();
            api.SubscribesTo<PaymentProcessedForOrder>();
            api.SubscribesTo<OrderCancelled>();
        }))
);
```

If types were then decorated e.g. using `System.ComponentModel.DescriptionAttribute` like this:
```csharp
using DescriptionAttribute = System.ComponentModel.DescriptionAttribute;

public record ProcessOrder(
    [Description("ID assigned to the order when created")] Guid OrderId,
    [Description("Items added to the order")] IReadOnlyList<OrderLine> Lines
);

public record OrderLine(
    [Description("The SKU of the added item")] string Sku,
    [Description("How many")] int Count
);

public record OrderProcessed(
    [Description("ID of the order that was processed")] Guid OrderId
);
```

then I believe this would be sufficient to generate a good description of the API using the AsyncAPI notation.

Moreover, Rebus' extensibility could be used to insert validations of incoming/outgoing messages to ensure that the API was not violated, this way guaranteeing that e.g. a Rebus instance would never publish anything that it had not declared beforehand.

Lastly, the `SubscribesTo<T>` declarations could be turned into automatic calls to `await bus.Subscribe<T>()` at bus startup, ensuring that the bus instance did in fact subscribe to the event types declared.

I'll definitely think some more about this. It would be interesting to hear you thoughts 🙂

**Comment by SbiCA** [comment]
I'm not using fleet manager I'm afraid. Nevertheless I can outline my envisioned scenarios (event tough I'm not an expert on asyncapi to be frank) 

 **docs**
Mainly I think this really valuable once you're passed a few simple messages similar to swagger ui, I'd like to be able to inspect a running (micro) service and see what it offers and how I can talk to it. 

Nice addition if I'm able to combine to generated docs into a bigger picture you could see how message `choreography` or `orchestration` is setup up including all you channel/transport concerns 🤯 This would be super nice architecture/learning property but of course requires quite some work I guess

**code generation**

Similar to points above her I see two valuable scenarios at them moment

Spec first => I receive a spec and would like to interact with a given service, with asyncapi I'd like to generate my ("client/server") code similar to what [grpc](https://docs.microsoft.com/en-us/aspnet/core/grpc/basics?view=aspnetcore-6.0) or [GraphQL](https://chillicream.com/docs/strawberryshake/get-started/console) allow you to do in .NET 

I really liked you're though on schemas and validation at this point you could of course also add in that part. In addition I think it would be invaluable for newcomers to already select the appropriate rebus Transport configuration e.g. for `MSSQL`, `RabitMQ` depending on the specs.

**Comment by SbiCA** [comment]
@ldeluigi yes I agree in addition what is a really interesting aspect could be performance since also .NET in general is now looking into AOT, which this type of integration would clearly allow for since I assume @mookid8000  has already some ideas with source generators 😄 https://devblogs.microsoft.com/dotnet/announcing-dotnet-8-preview-1/#native-aot

**Comment by SbiCA** [comment]
@mookid8000  just a crazy idea not sure how simple/hard would be to achieve this

why not using the same approach as minimal APIs is taking, that way it would be idiomatically very similar and IMO the gap between async/sync messaging would become smaller but maybe worth a separate issue since it's not really related to async apis?
```c#
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddRebus(
    configure => configure
        .Transport(t => t.UseAzureServiceBus(conn, "order_processor"))
        .Options(o => o.AsyncApi(api => {
            api.Title = "Order Processor";
            api.Version = new Version(2, 1, 0);
            api.Description = "Processes orders as they come in";
        }))
);


var app = builder.Build();


 // we handle commands of this type
app.Handles<ProcessOrder>( (order, bus) => {
    // code here ..
    bus.Publish(new OrderProcessed(...));
})
.Publishes<OrderProcessed>();


// we subscribe to these events
app.SubscribesTo<ShippingInfoAddedToOrder>(added => ...);
app.SubscribesTo<PaymentProcessedForOrder>(payment => ...);
app.SubscribesTo<OrderCancelled>(cancelation => .,..);

```

**Comment by mookid8000** [comment]
Can't see myself contributing to this tangent for now, so I'm closing the issue. Feel free to protest, if you do not approve.

**Comment by SbiCA** [comment]
Pity but no problem/relatable since it's quite a big change. Thx for all the hard work you put in the rebus @mookid8000
